chemprop/cli/hyperopt.py

```diff
@@ -1,16 +1,348 @@
+    "linked_hidden_size": tune.quniform(lower=300, upper=2400, q=100),
+    "max_lr": tune.loguniform(lower=1e-6, upper=1e-2),
+    "warmup_epochs": None
+}
+
+SEARCH_PARAM_KEYWORDS_MAP = {
+    "basic": ["depth", "ffn_num_layers", "dropout", "linked_hidden_size"],
+    "learning_rate": ["max_lr", "init_lr", "final_lr", "warmup_epochs"],
+    "all": list(AVAILABLE_SPACES.keys()),
+}
 
+def get_available_spaces(train_epochs: int) -> dict:
+
+    AVAILABLE_SPACES["warmup_epochs"] = tune.quniform(lower=1, upper=train_epochs // 2, q=1)
+
+    return AVAILABLE_SPACES
 
 class HyperoptSubcommand(Subcommand):
     COMMAND = "hyperopt"
     HELP = "perform hyperparameter optimization on the given task"
 
     @classmethod
     def add_args(cls, parser: ArgumentParser) -> ArgumentParser:
-        pass
+        parser = add_common_args(parser)
+        parser = add_train_args(parser)
+        return add_hyperopt_args(parser)
 
     @classmethod
     def func(cls, args: Namespace):
-        pass
+        args = process_common_args(args)
+        args = process_train_args(args)
+        args = process_hyperopt_args(args)
+        validate_common_args(args)
+        validate_train_args(args)
+        main(args)
+
+def add_hyperopt_args(parser: ArgumentParser) -> ArgumentParser:
+
+    hyperopt_args = parser.add_argument_group("Hyperparameter optimization arguments")
+
+    hyperopt_args.add_argument(
+        "--num-samples",
+        type=int,
+        default=10,
+        help="Number of hyperparameter optimization samples to run",
+    )
+
+    hyperopt_args.add_argument(
+        "--hyperopt-seed",
+        type=int,
+        default=0,
+        help="The initial seed used for choosing parameters in hyperparameter optimization trials. In each trial, the seed will be increased by one, skipping seeds previously used.",
+    )
+
+    hyperopt_args.add_argument(
+        "--hyperopt-save-dir",
+        type=Path,
+        help="Directory to save the hyperparameter optimization results",
+    )
+
+    hyperopt_args.add_argument(
+        "--hyperopt-checkpoint",
+        type=Path,
+        help="Path to a directory where hyperopt completed trial data is stored. Hyperopt job will include these trials if restarted. Can also be used to run multiple instances in parallel if they share the same checkpoint directory."
+    )
+
+    hyperopt_args.add_argument(
+        "--startup-random-iters",
+        type=int,
+        help="The initial number of trials that will be randomly specified before TPE algorithm is used to select the rest. By default will be half the total number of trials.",
+    )
+
+    hyperopt_args.add_argument(
+        "--manual-trial-dirs",
+        type=Path,
+        nargs="+",
+        help="Paths to save directories for manually trained models in the same search space as the hyperparameter search. Results will be considered as part of the trial history of the hyperparameter search.",
+    )
```

Comment:
This argument hasn't been used. I don't know if we can give manual trials within Ray Tune, and don't know if this functionality was popular in v1 though. If not many people use it and you think it is difficult to implement, I think we can consider the possibility of not supporting it.